src/StreamJsonRpc/Reflection/MessageFormatterRpcMarshaledContextTracker.cs

```diff
@@ -87,6 +90,50 @@ private enum MarshalMode
             MarshallingRealObject = 1,
         }
 
+        internal static (JsonRpcProxyOptions ProxyOptions, JsonRpcTargetOptions TargetOptions)? GetMarshalOptionsForType(Type type)
+        {
+            if (type.IsInterface is false)
+            {
+                return null;
+            }
+
+            return MarshaledTypes.GetOrAdd(type, (type) =>
+            {
+                foreach ((Type implicitlyMarshaledType, JsonRpcProxyOptions proxyOptions, JsonRpcTargetOptions targetOptions) in ImplicitlyMarshaledTypes)
+                {
+                    if (implicitlyMarshaledType == type ||
+                        (implicitlyMarshaledType.IsGenericTypeDefinition &&
+                         type.IsConstructedGenericType &&
+                         implicitlyMarshaledType == type.GetGenericTypeDefinition()))
+                    {
+                        return (proxyOptions, targetOptions);
+                    }
+                }
+
+                if (type.GetCustomAttribute<RpcMarshalableAttribute>() is not null)
+                {
+                    if (typeof(IDisposable).IsAssignableFrom(type) is false)
+                    {
+                        throw new NotSupportedException(Resources.MarshalableInterfaceNotDisposable);
+                    }
+
+                    if (type.GetEvents().Length > 0)
+                    {
+                        throw new NotSupportedException(Resources.MarshalableInterfaceHasEvents);
+                    }
+
+                    if (type.GetProperties().Length > 0)
+                    {
+                        throw new NotSupportedException(Resources.MarshalableInterfaceHasProperties);
+                    }
+
+                    return RpcMarshalableInterfaceDefaultOptions;
+                }
+
+                return null;
```

Comment:
@AArnott I added caching but should I cache `null`?
The only case in which we could have `null` here if I get an interface here but it is not marked as marshalable and it's not in the implicit list. Should I instead throw an exception if I ever encounter such interface? Or is there a valid use case (or a future use case) for an interface to be passed here that is not marshalable? (the only case I can think of is if there is a lower priority formatter that will also accept interfaces)

Reply:
Not caching `null` anymore as discussed in person.